### PR TITLE
[FLINK-33966] Fix the getNumRecordsInPerSecond Utility Function

### PR DESCRIPTION
**Pull request by Yang-LI-CS**
## What is the purpose of the change

Fix the getNumRecordsInPerSecond Utility Function to addresses the redundant check and ensures correct metric fetching for non-source operators.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changes to the `CustomResourceDescriptors`: no
  - Core observer or reconciler logic that is regularly executed: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-autoscaler/src/main/java/org/apache/flink/autoscaler/metrics/ScalingMetrics.java [hunk 1]
```diff
@@ -221,7 +221,7 @@ private static double getNumRecordsInPerSecond(
             numRecordsInPerSecond =
                     flinkMetrics.get(FlinkMetric.SOURCE_TASK_NUM_RECORDS_IN_PER_SEC);
         }
-        if (isSource && (numRecordsInPerSecond == null || numRecordsInPerSecond.getSum() == 0)) {
+        if (!isSource && (numRecordsInPerSecond == null || numRecordsInPerSecond.getSum() == 0)) {
             numRecordsInPerSecond =
                     flinkMetrics.get(FlinkMetric.SOURCE_TASK_NUM_RECORDS_OUT_PER_SEC);
         }
```
Comment (1996fanrui) [inline flink-autoscaler/src/main/java/org/apache/flink/autoscaler/metrics/ScalingMetrics.java:227]: Thanks @Yang-LI-CS for the PR!

These 2 conditions are indeed the same, it maybe unexpected intuitively. Maybe here logic is numRecordsInPerSecond from 3 metrics:

1. FlinkMetric.NUM_RECORDS_IN_PER_SEC
2. FlinkMetric.SOURCE_TASK_NUM_RECORDS_IN_PER_SEC
3. FlinkMetric.SOURCE_TASK_NUM_RECORDS_OUT_PER_SEC

When 1 is unavailable, get it from 2. When 2 is unavailable, get it from 3.

I'm not sure about it. Need to confirm with @mxm . 

- If my guess is right, it's not a bug. And I suggest this PR add a little comments to clarify it. 
- If my guess is wrong, it's a bug.  We can fix it.

WDYT?

Comment (1996fanrui) [reply flink-autoscaler/src/main/java/org/apache/flink/autoscaler/metrics/ScalingMetrics.java:227]: By the way, current change doesn't make sense.

When a task isn't source, it doesn't have `SOURCE_TASK_NUM_RECORDS_OUT_PER_SEC` metric.

Comment (Yang-LI-CS) [reply flink-autoscaler/src/main/java/org/apache/flink/autoscaler/metrics/ScalingMetrics.java:227]: @1996fanrui thanks, I'll add some comments and do as your suggestion 👍

File: flink-autoscaler/src/main/java/org/apache/flink/autoscaler/metrics/ScalingMetrics.java [hunk 2]
```diff
@@ -217,14 +217,13 @@ private static double getNumRecordsInPerSecond(
             JobVertexID jobVertexID,
             boolean isSource) {
         var numRecordsInPerSecond = flinkMetrics.get(FlinkMetric.NUM_RECORDS_IN_PER_SEC);
-        if (isSource && (numRecordsInPerSecond == null || numRecordsInPerSecond.getSum() == 0)) {
-            numRecordsInPerSecond =
-                    flinkMetrics.get(FlinkMetric.SOURCE_TASK_NUM_RECORDS_IN_PER_SEC);
-        }
```
Comment (1996fanrui) [inline flink-autoscaler/src/main/java/org/apache/flink/autoscaler/metrics/ScalingMetrics.java:223]: Why the `SOURCE_TASK_NUM_RECORDS_IN_PER_SEC` part is removed?

File: flink-autoscaler/src/main/java/org/apache/flink/autoscaler/metrics/ScalingMetrics.java (hunk 1 above)
Comment (Yang-LI-CS) [reply flink-autoscaler/src/main/java/org/apache/flink/autoscaler/metrics/ScalingMetrics.java]: @1996fanrui This time I have just removed the unused if condition and add 1 comment.

To be sure, let's wait feedback from @mxm

File: flink-autoscaler/src/main/java/org/apache/flink/autoscaler/metrics/ScalingMetrics.java (hunk 2 above)
Comment (Yang-LI-CS) [reply flink-autoscaler/src/main/java/org/apache/flink/autoscaler/metrics/ScalingMetrics.java]: @1996fanrui  Sry, I have put it back and just add the comments

File: flink-autoscaler/src/main/java/org/apache/flink/autoscaler/metrics/ScalingMetrics.java (hunk 1 above)
Comment (Yang-LI-CS) [reply flink-autoscaler/src/main/java/org/apache/flink/autoscaler/metrics/ScalingMetrics.java]: I have just add some comments to clarify this at last 🙏

File: flink-autoscaler/src/main/java/org/apache/flink/autoscaler/metrics/ScalingMetrics.java [hunk 3]
```diff
@@ -216,15 +216,20 @@ private static double getNumRecordsInPerSecond(
...
         if (isSource && (numRecordsInPerSecond == null || numRecordsInPerSecond.getSum() == 0)) {
             numRecordsInPerSecond =
                     flinkMetrics.get(FlinkMetric.SOURCE_TASK_NUM_RECORDS_IN_PER_SEC);
         }
+        // If the vertex is the source and SOURCE_TASK_NUM_RECORDS_IN_PER_SEC metric not available
+        // then use SOURCE_TASK_NUM_RECORDS_OUT_PER_SEC metric
         if (isSource && (numRecordsInPerSecond == null || numRecordsInPerSecond.getSum() == 0)) {
             numRecordsInPerSecond =
                     flinkMetrics.get(FlinkMetric.SOURCE_TASK_NUM_RECORDS_OUT_PER_SEC);
         }
```
Comment (1996fanrui) [inline flink-autoscaler/src/main/java/org/apache/flink/autoscaler/metrics/ScalingMetrics.java:231]:
```suggestion
        // Generate numRecordsInPerSecond from 3 metrics:
        // 1. If the vertex is not the source, use NUM_RECORDS_IN_PER_SEC metric
        var numRecordsInPerSecond = flinkMetrics.get(FlinkMetric.NUM_RECORDS_IN_PER_SEC);
        // 2. If the vertex is the source, use SOURCE_TASK_NUM_RECORDS_IN_PER_SEC metric first.
        if (isSource && (numRecordsInPerSecond == null || numRecordsInPerSecond.getSum() == 0)) {
            numRecordsInPerSecond =
                    flinkMetrics.get(FlinkMetric.SOURCE_TASK_NUM_RECORDS_IN_PER_SEC);
        }
        // 3. If the vertex is the source and SOURCE_TASK_NUM_RECORDS_IN_PER_SEC metric not available
        // then use SOURCE_TASK_NUM_RECORDS_OUT_PER_SEC metric
        if (isSource && (numRecordsInPerSecond == null || numRecordsInPerSecond.getSum() == 0)) {
            numRecordsInPerSecond =
                    flinkMetrics.get(FlinkMetric.SOURCE_TASK_NUM_RECORDS_OUT_PER_SEC);
        }
```

How about adding the `Generate numRecordsInPerSecond from 3 metrics:` and the `1. 2. 3.`?

Comment (Yang-LI-CS) [reply flink-autoscaler/src/main/java/org/apache/flink/autoscaler/metrics/ScalingMetrics.java]: Good idea! 🙏

**Review by mxm — COMMENTED** [review]
Looks good! I think this logic needs some comments because it is hard to understand the reasoning behind the logic. I added some additional suggestions.

**Comment by Yang-LI-CS** [comment]
> Looks good! I think this logic needs some comments because it is hard to understand the reasoning behind the logic. I added some additional suggestions.

Thanks @mxm I have added your suggestion :)

**Review by mxm — APPROVED** [review]

File: flink-autoscaler/src/main/java/org/apache/flink/autoscaler/metrics/ScalingMetrics.java [hunk 4]
```diff
@@ -216,15 +216,21 @@ private static double getNumRecordsInPerSecond(
             Map<FlinkMetric, AggregatedMetric> flinkMetrics,
             JobVertexID jobVertexID,
             boolean isSource) {
+        // Generate numRecordsInPerSecond from 3 metrics:
+        // 1. If available, directly use the NUM_RECORDS_IN_PER_SEC task metric.
         var numRecordsInPerSecond = flinkMetrics.get(FlinkMetric.NUM_RECORDS_IN_PER_SEC);
+        // 2. If the former is unavailable and the vertex contains a source operator, use the corresponding source operator metric.
```
Comment (mxm) [inline flink-autoscaler/src/main/java/org/apache/flink/autoscaler/metrics/ScalingMetrics.java:222]: We might have to run `mvn spotless:apply`.

File: flink-autoscaler/src/main/java/org/apache/flink/autoscaler/metrics/ScalingMetrics.java [hunk 5]
```diff
@@ -216,15 +216,23 @@ private static double getNumRecordsInPerSecond(
...
         var numRecordsInPerSecond = flinkMetrics.get(FlinkMetric.NUM_RECORDS_IN_PER_SEC);
+        // 2. If the former is unavailable and the vertex contains a source operator, use the
+        // corresponding source operator metric.
         if (isSource && (numRecordsInPerSecond == null || numRecordsInPerSecond.getSum() == 0)) {
             numRecordsInPerSecond =
                     flinkMetrics.get(FlinkMetric.SOURCE_TASK_NUM_RECORDS_IN_PER_SEC);
         }
+        // 3. If the vertex contains a source operator which does not emit input metrics, use output
+        // metrics instead.
+        // then use SOURCE_TASK_NUM_RECORDS_OUT_PER_SEC metric
```
Comment (1996fanrui) [inline flink-autoscaler/src/main/java/org/apache/flink/autoscaler/metrics/ScalingMetrics.java:230]:
```suggestion
```

Is this leftover?

Comment (Yang-LI-CS) [reply flink-autoscaler/src/main/java/org/apache/flink/autoscaler/metrics/ScalingMetrics.java:230]: @1996fanrui indeed, I'll remove it

**Review by 1996fanrui — APPROVED** [review]
Thanks @Yang-LI-CS for the contribution!

LGTM assuming the CI is green.

**Pull request merged**